plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,401 @@
+				struct ukplat_lcpu_func, lentry))
+			break;
+	}
+
+	intctrl_ack_irq(UKPLAT_ARM64_IPI_RUN_IRQ);
+
+	return 1;
+}
+
+int smp_init(void)
+{
+	int fdt_cpu;
+	const fdt32_t *naddr_prop, *size_prop, *id_reg;
+	const struct fdt_property *cpu_prop;
+	int naddr, size = 0, cell;
+	unsigned int cpu_idx;
+	int subnode;
+
+	/* Init the info for the bootstraping cpu */
+	bsp_cpu_id = ukplat_lcpu_id();
+	cpus[0].id = bsp_cpu_id;
+	cpus[0].state = LCPU_STATE_BUSY;
+	ukarch_spin_init(&cpus[0].lock);
+	uk_pr_info("Bootstrapping processor has the ID %d\n", bsp_cpu_id);
+
+	/* Init the secondary cpus vector */
+	for (cpu_idx = 1; cpu_idx < CONFIG_UKPLAT_LCPU_MAXCOUNT; cpu_idx++) {
+		cpus[cpu_idx].id = (__u32) -1;
+		cpus[cpu_idx].state = LCPU_STATE_OFFLINE;
+	}
+	smp_numcores = 0;
+
+	/* Search for cpus in DTB */
+	fdt_cpu = fdt_path_offset(_libkvmplat_cfg.dtb, "/cpus");
+	if (fdt_cpu < 0) {
+		uk_pr_err("cpus node is not found in device tree\n");
+		return -1;
+	}
+
+	/* Get address cells */
+	naddr_prop = fdt_getprop(_libkvmplat_cfg.dtb,
+					fdt_cpu, "#address-cells", &cell);
+	if (naddr_prop == NULL) {
+		naddr = FDT_ADDR_CELLS_DEFAULT;
+		uk_pr_warn("Using default value 2 for cpu address-cells.\n");
+	} else {
+		naddr = fdt32_to_cpu(naddr_prop[0]);
+	}
+	if (naddr < 0 || naddr >= FDT_MAX_NCELLS) {
+		uk_pr_err("Invalid adress-cells size!\n");
+		return -1;
+	}
+
+	/* Get size cells */
+	size_prop = fdt_getprop(_libkvmplat_cfg.dtb, fdt_cpu,
+			"#size-cells", &cell);
+	if (size_prop == NULL) {
+		naddr = FDT_SIZE_CELLS_DEFAULT;
+		uk_pr_warn("Using default value of 0 for cpu size-cells.\n");
+	} else {
+		size = fdt32_to_cpu(size_prop[0]);
+	}
+	if (size != 0) {
+		uk_pr_err("Invalid size-cells size for cpu node!\n");
+		return -1;
+	}
+
+	/* Search all the cpu nodes in DTB */
+	cpu_idx = 0;
+	fdt_for_each_subnode(subnode, _libkvmplat_cfg.dtb, fdt_cpu) {
+		int prop_len = 0;
+		__u32 cpu_id;
+
+		cpu_prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"device_type", &prop_len);
+		if (!cpu_prop)
+			continue;
+		if (strcmp(cpu_prop->data, "cpu"))
+			continue;
+
+		cpu_prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"enable-method", NULL);
+		if (!cpu_prop) {
+			uk_pr_err("enable-method field is missing \
+				in cpu fdt node!\n");
+			continue;
+		} else if (strcmp(cpu_prop->data, "psci")) {
+			uk_pr_err("enable-method is not PSCI in \
+				cpu fdt node: (%s)\n",
+				cpu_prop->data);
+			continue;
+		}
+
+		cpu_prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"reg", &prop_len);
+		if (cpu_prop == NULL || prop_len <= 0) {
+			uk_pr_err("Error when reading cpu reg field \
+				(cpu id) in fdt!\n");
+			continue;
+		}
+
+		id_reg = (const fdt32_t *) cpu_prop->data;
+		cpu_id = fdt_reg_read_number(id_reg, naddr);
+		if (cpu_id != bsp_cpu_id) {
+			cpu_idx++;
+			cpus[cpu_idx].id = cpu_id;
+			ukarch_spin_init(&cpus[cpu_idx].lock);
+			uk_pr_info("Initialized core (%u)\n",
+				cpus[cpu_idx].id);
+		}
```

Comment:
I will add this check for now, but I think this could be a greater issue. If the ids are broken in the device tree, this may generate problems in the API functions, when we are trying to associate the current cpu with it's `arm64_cpu` structure (e.g. when calling `ukplat_lcpu_idx()`).
Maybe it would be a good solution to add a second check in `ukplat_lcpu_idx()`? I am thinking about something like giving an error message if the core's physical id from MPIDR can't be found in the cpus array.

Reply:
I agree. Also, I think this can be a greater issue and we should add a check to `ukplat_lcpu_idx()` too (see my comment on that function).

Reply:
Exactly, this can happen in any core, but for cores other than the boot-core, as you pointed out, that will be caught by `ukplat_lcpu_idx()`.

To check if the bootcore id is in the dtb, a more elegant way than defining a `bsp_found` variable would be to:
* Remove the initialization of the bootcore (ie `cpus[0]) at the beginning of the function.
* In your loop, do not differentiate between the bootcore and secondary cores. Instead initialize all members of `cpus[]` from what you read in the dtb.
* After your loop, call `ukplat_lcpu_idx()` - will fail if the bootcore is not in the dts - and set the state of the bootcore to BUSY.

```
this_cpu_idx = ukplat_lcpu_idx();
if (this_cpu_idx == -1)
        return -1;
cpus[this_cpu_idx].state = LCPU_STATE_BUSY;
```